### DISCUSSION TITLE
Pass api routes html and image to component render

### DISCUSSION THREAD

**Discussion by 2gud4u** [vercel/next.js#38337] [Help] [answered]
Hello community.-

As i was taught at this post of mine https://github.com/vercel/next.js/discussions/38309 i have succesfully called my api routes but at component render i have a next image error this is my code.-

```
const [education, setEducation] = useState(null);
const [logo, setLogo] = useState(null);
const [isLoadingEducation, setLoadingEducation] = useState(false)
const [isLoadingLogo, setLoadingLogo] = useState(false)



useEffect(() => {
        if (redirectSeconds == 5) {
            fetch('/api/education')
                .then((res) => res.json())
                .then((education) => {
                    setEducation(education)
                    setLoadingEducation(true)
                })
            
            console.log('4 seconds heading')
            //return
        }else if(redirectSeconds == 3){
            fetch('/api/logo')
                .then((res) => res.json())
                .then((logo) => {
                    setLogo(logo)
                    setLoadingLogo(true)
                })
        }else if(redirectSeconds == 0){
            router.push('/Subscription');
            console.log('0 seconds pricing subscriptions')
        }


        setTimeout(() => {
            console.log(redirectSeconds);
            setRedirectSeconds((redirectSeconds) => redirectSeconds - 1);
        }, 1000)


}, [redirectSeconds]);

```
Then at the component i want to display the server html.-

```
return (
            <div id='animation' style={{display:'flex', flexDirection: 'row', zIndex: -100, justifyContent: 'center', alignItems: 'center', height: '100%', width: '100%', margin: '0 0 3rem'}}>
            <div id='animationcat' style={{position: 'absolute', zIndex: 100, left: '45%', justifyContent: 'center', alignItems: 'center', height: '750px', width: '980px'}}></div>
            if(isLoadingEducation)
            {
                <h1>{education}</h1>
            }
            if(isLoadingLogo)
            {
                <Image src={logo} alt="logo" width="100%" height='100%' />
            }
            </div>
        );
}

```

I get this error on component run TypeError: Cannot read properties of null (reading 'default') at image.js

**Comment by icyJoseph** [comment, marked as answer]
Hi,

That JSX syntax is not valid, further down I explore a bit more into that.

What's happening here is that since `logo` is `null` initially, then the Image is trying to read from that. Because the JSX you have is invalid the `if` condition is doing nothing. 

Here's an expansion into that:

## Rendering conditionally

JSX is a flavour of JavaScript that can intake expressions:

```jsx
{isOpen ? <Open /> : <Closed />}
```

For example the above, is valid JSX, and so is:

```jsx
<div>
    {console.log("Wow don't do this in production")}
    <p>Hi there</p>
</div>
```

There's no pretty way to inline `if-else` statements, other than short-circuits, ternaries, partial render functions, or a very much NOT recommended IIFE way (only useful to show off at parties).

### Short-circuit

```jsx
{isValid && <p>valid</p>}
```

As long as `isValid` is empty string, false, null, then nothing is rendered. If `isValid` is `0` number, then `0` is rendered. For all other truthy values, the `&&` knows that the value of `isValid && <p>valid</p>` depends on the second input, so that is rendered.

Short-circuiting is a very nice programming concept that works well in JavaScript.

### Ternaries

```jsx
{isOpen ? <Open /> : <Closed />}
```

### Partial render

I personally avoid this... and refactor them into components of their own.

```jsx
  const renderSomething = () => {
    if (something) return <span></span>;
    return <div></div>;
  };

  return (
    <div>
      <h1>Hello</h1>
      {renderSomething()}
    </div>
  );
```

### IIFE (avoid)

Immediately Invoked Function Expression

```jsx
// avoid this... ❗❗❗❗❗
 <>
        {(() => { 
          if (something) {
            return <span></span>;
          }
          return <div></div>;
        })()}
  </>
```

### Booleans for loading states

I'd say, avoid `boolean`'s for loading states as well, and use strings that describe in which step of the process you are, such as `'pending'`, `'resolved'`, `'rejected'`. More on that here, https://kentcdodds.com/blog/stop-using-isloading-booleans. As always, take articles and opinions like this with a grain of salt.

## Effects

I also notice you've packed quite a `useEffect` in there. It is a good practice to try, if possible, to divide into as many useEffects as possible. As long as the effects can be separate of course.

It is also good to get used to cleaning up effects, by returning a clean up function.

```ts
  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    if (redirectSeconds == 5) {
      fetch("/api/education", { signal })
        .then((res) => res.json())
        .then((education) => {
          setEducation(education);
          setLoadingEducation(true);
        })
        .catch((e) => {
          if (signal.aborted) console.log("Request was cancelled");
          // do something else with the error
        });

      console.log("4 seconds heading");
      //return
    } else if (redirectSeconds == 3) {
      fetch("/api/logo", { signal })
        .then((res) => res.json())
        .then((logo) => {
          setLogo(logo);
          setLoadingLogo(true);
        })
        .catch((e) => {
          if (signal.aborted) console.log("Request was cancelled");
          // do something else with the error
        });
    } else if (redirectSeconds == 0) {
      router.push("/Subscription");
      console.log("0 seconds pricing subscriptions");
    }

    const timeout = setTimeout(() => {
      console.log(redirectSeconds);
      // here, it is best to change the, arbitrary name, given to this variable
      // just because it is easier to read
      setRedirectSeconds((prev) => prev - 1);
    }, 1000);

    return () => {
      controller.abort();
      clearTimeout(timeout);
    };
  }, [redirectSeconds, router]);
```

A lot going on there, first and foremost, if the `router.push` results in this component being unmounted, then we are good. Otherwise, we might end up in a runaway effect.

Second, clearing the timeout, in case this effect dependencies change.

Third, cancelling the fetch requests with an abort controller, in case the dependencies change. 

In both, second and third case, we don't care about the async work we started anymore, because our variables have changed. If we don't stop it, it might resolve, in the case of fetch, later than the newly started request, and it'll override our state! In the case of timers, the results might even be stranger, but one thing that's for sure is that, if the component un-mounts, and the timer runs its callback, there'll be an attempt to set state on a component that's no longer there, AKA, memory leak.

**Reply by 2gud4u** [reply]
Wow i dont know if this will solve my problem but it address to many concepts i didnt have clear, again thank you for everything !!!!

**Reply by icyJoseph** [reply]
Let us know how it goes! Glad to help out!

**Reply by 2gud4u** [reply]
Yes, this is the answer !! It only took me more time to display correctly the Image, i was going to mark as answer, but somebody already did. Thank you again @icyJoseph  !!!! I was with this problem for 6 days
